ethexe/contracts/src/Router.sol

```diff
@@ -379,36 +380,59 @@ contract Router is IRouter, OwnableUpgradeable, ReentrancyGuardTransient {
 
     /* Helper private functions */
 
-    function _createProgram(bytes32 _codeId, bytes32 _salt) private returns (address) {
-        Storage storage router = _router();
+    function _createProgram(bytes32 _codeId, Storage storage router) private returns (address program) {
         require(router.genesisBlock.hash != bytes32(0), "router genesis is zero; call `lookupGenesisHash()` first");
 
         require(
             router.protocolData.codes[_codeId] == Gear.CodeState.Validated,
             "code must be validated before program creation"
         );
 
-        // Check for duplicate isn't necessary, because `Clones.cloneDeterministic`
-        // reverts execution in case of address is already taken.
-        address actorId =
-            Clones.cloneDeterministic(router.implAddresses.mirrorProxy, keccak256(abi.encodePacked(_codeId, _salt)));
+        address _mirrorAbi = router.implAddresses.mirrorAbi;
 
-        router.protocolData.programs[actorId] = _codeId;
-        router.protocolData.programsCount++;
+        assembly {
+            let contractSize := extcodesize(_mirrorAbi)
 
-        emit ProgramCreated(actorId, _codeId);
+            let ctr := mload(0x40)
+            mstore(ctr, shl(0xa8, 0x6080806040523460155761))
+            let offset := 0x0b
+            mstore8(add(ctr, offset), byte(0x1e, contractSize))
+            mstore8(add(ctr, add(offset, 0x01)), byte(0x1f, contractSize))
+            offset := add(offset, 0x02)
+            mstore(add(ctr, offset), shl(0xa0, 0x90816100198239f35b5f80fd))
+            offset := add(offset, 0x0c)
+            mstore(0x40, add(ctr, offset))
 
-        return actorId;
-    }
+            let prefixSize := offset
+            let bytecodeSize := extcodesize(_mirrorAbi)
+            let fullBytecodeSize := add(prefixSize, bytecodeSize)
 
-    function _createDecoder(address _implementation, bytes32 _salt, address _mirror) private returns (address) {
-        address decoder = Clones.cloneDeterministic(_implementation, _salt);
+            let code := mload(0x40)
 
-        IMirrorDecoder(decoder).initialize(_mirror);
+            mstore(code, fullBytecodeSize)
 
-        return decoder;
+            let src := mload(ctr)
+            let dest := add(code, 0x20)
+            for { let i := 0 } lt (i, prefixSize) { i:= add(i, 1) } {
+                mstore8(add(dest, i), byte(i, src))
+            }
+
+            extcodecopy(_mirrorAbi, add(add(code, 0x20), prefixSize), 0, bytecodeSize)
+
+            program := create(0, add(code, 0x20), mload(code))
```

Comment:
previously we were just deployed ERC1167, which essentially created proxy contract of 45 bytes, but now we are copying MirrorAbi's bytecode (2977 bytes). now deploying new programs will probably be much more expensive.